### ISSUE TITLE
Clang assertation

### ISSUE THREAD

**Issue by objwyh** [llvm/llvm-project#58389]

**Comment by shafik** [comment]
CC @tahonermann Could this be related to you fix for https://github.com/llvm/llvm-project/issues/57791

**Comment by tahonermann** [comment]
@shafik, I don’t think so. The changes made for issue #57791 are specific to the Microsoft style inline assembly enabled with the `-fasm-blocks` option. I can take a closer look tomorrow to make sure though.

**Comment by nicolas17** [comment]
clang 10.0.0 also fails with an assertion, so this is not due to a recent change.

```
clang++: /home/nicolas/llvm-project/llvm/lib/IR/InlineAsm.cpp:39: llvm::InlineAsm::InlineAsm(llvm::FunctionType *, const std::string &, const std::string &, bool, bool, llvm::InlineAsm::AsmDialect): Assertion `Verify(getFunctionType(), constraints) && "Function type not legal for constraints!"' failed.
...
clang-10: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 10.0.0 (https://github.com/llvm/llvm-project d32170dbd5b0d54436537b6b75beaf44324e0c28)
Target: aarch64-unknown-linux-gnu
Thread model: posix
```

**Comment by tahonermann** [comment]
I spent a little time debugging this.

First, I had to research the semantics of a `"*"` constraint. I wasn't able to find documentation, but the `TargetInfo::validateInputConstraint()` in `lib/Basic/TargetInfo.cpp` indicates that it means that any register can be used (https://github.com/llvm/llvm-project/blob/release/15.x/clang/lib/Basic/TargetInfo.cpp#L836-L837). Changing the constraint to `"r"` sufficed to avoid the problem. Changing it to "g" produced an appropriate error.

Second, I determined that the issue is not dependent on use of a bit-field; the problem reproduces with this example:
```
struct S {
  int dm;
} s;
void f() {
  __asm__ volatile("movl %%eax, %0" :: "*"(s.dm));
}
```

The assertion failure occurs at line 39 of `llvm::InlineAsm::InlineAsm()` as shown below:
```
llvm/lib/IR/InlineAsm.cpp:
 30 InlineAsm::InlineAsm(FunctionType *FTy, const std::string &asmString,
 31                      const std::string &constraints, bool hasSideEffects,
 32                      bool isAlignStack, AsmDialect asmDialect, bool canThrow)
 33     : Value(PointerType::getUnqual(FTy), Value::InlineAsmVal),
 34       AsmString(asmString), Constraints(constraints), FTy(FTy),
 35       HasSideEffects(hasSideEffects), IsAlignStack(isAlignStack),
 36       Dialect(asmDialect), CanThrow(canThrow) {
 37 #ifndef NDEBUG
 38   // Do various checks on the constraint string and type.
 39   cantFail(verify(getFunctionType(), constraints));
 40 #endif
 41 }
```

`llvm::InlineAsm::verify()` returns in error after having determined that the number of inputs (0) doesn't match the number of expected parameters (1) for the `Ty` parameter.
```
llvm/lib/IR/InlineAsm.cpp
260 Error InlineAsm::verify(FunctionType *Ty, StringRef ConstStr) {
...
322   if (Ty->getNumParams() != NumInputs)
323     return makeStringError("number of input constraints does not match number "
324                            "of parameters");
...
328 }
```

The arguments passed are:
```
(gdb) p Ty->dump()
void (i32)
.....
(gdb) p ConstStr
$2 = {
  static npos = 18446744073709551615,
  Data = 0x14cd33b0 "~{dirflag},~{fpsr},~{flags}",
  Length = 27
}
```

The miscalculation of the number of inputs occurs because `ConstStr` doesn't include a constraint for the input variable (the `~{dirflag}`, `~{fpsr}`, and `~{flags}` all correspond to implicit machine dependent clobber constraints). The input constraint goes missing because the call to `SimplifyConstraint()` at line 2558 of `CodeGenFunction::EmitAsmStmt()` below simplifies the constraint from `"*"` to `""`. Since no output variables are specified, `Constraints` is empty at this point and the append of `InputConstraint` to it at line 2613 ends up having no effect.
```
clang/lib/CodeGen/CGStmt.cpp:
2328 void CodeGenFunction::EmitAsmStmt(const AsmStmt &S) {
....
2545   for (unsigned i = 0, e = S.getNumInputs(); i != e; i++) {
....
2557     std::string InputConstraint(S.getInputConstraint(i));
2558     InputConstraint = SimplifyConstraint(InputConstraint.c_str(), getTarget(),
2559                                          &OutputConstraintInfos);
....
2613     Constraints += InputConstraint;
2614   }
....

Since `Constraints` remains empty, no comma is appended before the clobber constraints are appended; the input constraint is therefore not reflected in the list of constraints.

`SimplifyConstraint()` is explicit in its intention to ignore `"*"` (and some other) constraints, but its method of ignoring them has the effect of breaking the constraint validation performed by `llvm::InlineAsm::verify()`.
```
clang/lib/CodeGen/CGStmt.cpp:
2092 static std::string
2093 SimplifyConstraint(const char *Constraint, const TargetInfo &Target,
2094                  SmallVectorImpl<TargetInfo::ConstraintInfo> *OutCons=nullptr) {
2095   std::string Result;
2096
2097   while (*Constraint) {
2098     switch (*Constraint) {
....
2102     // Ignore these
2103     case '*':
2104     case '?':
2105     case '!':
2106     case '=': // Will see this and the following in mult-alt constraints.
2107     case '+':
2108       break;
....
2134     }
2135
2136     Constraint++;
2137   }
2138
2139   return Result;
2140 }
```

**Comment by tahonermann** [comment]
This issue appears to be strongly related to https://github.com/llvm/llvm-project/issues/50219.

**Comment by tahonermann** [comment]
@nikic @xiangzh1 given some of your past work in this area, do you have any thoughts on how to address this? Please see the analysis in [the previous comment](https://github.com/llvm/llvm-project/issues/58389#issuecomment-1296015624).

**Comment by xiangzh1** [comment]
Very good/careful analysis!
So the key problem is why SimplifyConstraint() ignore "*" . I think the reason is that current llvm didn't implement the "*" constraint for an independent constraint (it mainly use with other char together, for example "*m" to imply indirection). 
I'd like to come the the problem self: do we need to support the independent constraint "*" ? Why we not write "r" in the source code.

**Comment by tahonermann** [comment]
@xiangzh1:
> I'd like to come the the problem self: do we need to support the independent constraint " * " ? Why we not write " r " in the source code.

Good questions that I don't know the answer to.

gcc and icc both reject the original test case. Here is gcc's diagnostic:
```
<source>: In function 'void f()':
<source>:5:3: warning: 'asm' operand 0 probably does not match constraints
    5 |   __asm__ volatile("movl %%eax, %0" :: "*"(s.dm));
      |   ^~~~~~~
<source>:5:3: error: impossible constraint in 'asm'
```
Here is icc's diagnostic:
```
<source>(5): catastrophic error: Cannot match asm operand constraint
compilation aborted for <source> (code 1)
```

The same results occur if the constraint is changed from `"*"` to `"?"` or `"!"`. I guess Clang should be diagnosing the constraint? I'm afraid I'm not sufficiently familiar with these constraints to know what the restrictions for their use are. Pairing them with "r" results in all three compilers accepting the test case.

**Comment by xiangzh1** [comment]
It's normal that you're not familiar with it, because it very rarely used. I +1 for it should be sync with gcc (diagnosing these independent constraints as invalid). For these special constraints, only we meet strong requirement from users we should implement them.
